persistence-api/src/main/java/io/stargate/db/limiter/AsyncRateLimiter.java

```diff
@@ -0,0 +1,218 @@
+
+  /**
+   * Threshold under which a task is executed/enqueued immediately instead of being scheduled. Set
+   * to 1 millisecond because the typical scheduling resolution is in milliseconds.
+   */
+  private final long schedulingThresholdNanos;
+
+  /**
+   * The minimum time that has to elapse for a permit to be granted, in nanoseconds. In other words,
+   * the inverse of the permitted rate.
+   */
+  private volatile double nanosPerPermit;
+
+  /**
+   * The amount of time we are allowed to have "in reserve", i.e. elapsed time that we can use for
+   * new permits. This is limited to make sure that we don't run unlimited if the process has not
+   * been active for some time, or has not needed all of the provided rate.
+   */
+  private final long reserveWindowNanos;
+
+  /**
+   * The time to which permits have already been allocated. Every allocation increases this by
+   * {@code permits * nanosPerPermits}, and the associated operation is scheduled for the resulting
+   * time (if it is in the future). If we are running at capacity this would be in the future, and
+   * if we are not this may be in the past. If it gets more than reserveWindowNanos nanoseconds
+   * behind the current time, the next allocation will adjust it to make sure we do not allow long
+   * unlimited runs due to long periods of inactivity.
+   */
+  private final AtomicLong consumedToTime;
+
+  /**
+   * Constructs a limiter.
+   *
+   * @param executor the executor on which tasks that are rate limited are scheduled.
+   * @param rate the number of permits to allow per rateUnit.
+   * @param rateUnit time unit for {@code rate}.
+   * @param reserveWindow the amount of time we are allowed keep in reserve for work that is not yet
+   *     claimed; this should be several times the time a unit of work is expected to take, so that
+   *     the job cannot lose permits because of long-running tasks while it is operating at
+   *     capacity.
+   * @param reserveWindowUnit time unit for the reserve window
+   * @param schedulingThreshold a time under which a task is executed directly instead of being
+   *     scheduler. Meaning, if when {@link #acquireAndExecute} is called, the time before execution
+   *     is computed to a value below this threshold, then the task is executed directly instead of
+   *     being scheduled on the executor. This is meant to save the cost of scheduling for very
+   *     small delays.
+   * @param schedulingThresholdUnit time unit for schedulingThreshold.
+   */
+  public AsyncRateLimiter(
+      ScheduledExecutorService executor,
+      long rate,
+      TimeUnit rateUnit,
+      long reserveWindow,
+      TimeUnit reserveWindowUnit,
+      long schedulingThreshold,
+      TimeUnit schedulingThresholdUnit) {
+    this.executor = executor;
+    this.nanosPerPermit = rateUnit.toNanos(1) * 1.0 / rate;
+    this.reserveWindowNanos = reserveWindowUnit.toNanos(reserveWindow);
+    this.consumedToTime = new AtomicLong(System.nanoTime());
```

Comment:
I've left it out for now because the unit test don't use it, and if we really want to allow some time injection, we'd also need to take care of the `System.nanoTime` in `acquireAndExecute` and I'm not sure about adding that complexity if it's unused. This can obviously be changed later.